docs(examples): hello world, prompt

rajbos opened the pull request:
See the conversation in #83 

- Adding `examples` folder with easy to start projects (see `README` in that folder)
- Example 1: hello world
- Example 2: handle the user prompt by sending it to the Copilot API and returning the result

gr2m commented on examples/1-hello-world/package.json:
```diff
@@ -0,0 +1,20 @@
+  },
+  "scripts": {
+    "start": "node index.js",
+    "watch": "node --watch index.js"
```
does node support a `--watch` flag now?

rajbos replied:
Yep

gr2m replied:
woah today I learned, thanks! 
https://nodejs.org/api/cli.html#--watch

rajbos replied:
Copilot told me it was available!

gr2m commented on examples/1-hello-world/package.json:
```diff
@@ -0,0 +1,20 @@
+  "author": "",
+  "license": "ISC",
+  "dependencies": {
+    "@copilot-extensions/preview-sdk": "^4.0.3"
```
I would use latest as [next.js does](https://github.com/vercel/next.js/blob/874454f3fab4cf332a8436ad38497cfc4ab91219/examples/hello-world/package.json#L9)

```suggestion
    "@copilot-extensions/preview-sdk": "latest"
```

gr2m replied:
we could also set it to a relative path instead, as the source code can be imported directly, we don't have a build step

```suggestion
    "@copilot-extensions/preview-sdk": "../../index.js"
```

gr2m commented on examples/1-hello-world/package.json:
```diff
@@ -0,0 +1,20 @@
+{
+  "name": "hello-world",
+  "version": "0.0.1",
+  "description": "Hello world example",
+  "main": "index.js",
+  "type": "module",
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/copilot-extensions"
+  },
```
I would remove all unnecessary keys, but add private: true

```suggestion
  "private": true,
```

gr2m commented on examples/1-hello-world/package.json:
```diff
@@ -0,0 +1,20 @@
+    "start": "node index.js",
+    "watch": "node --watch index.js"
+  },
+  "author": "",
+  "license": "ISC",
```
```suggestion
```

gr2m commented on examples/1-hello-world/package-lock.json:
not sure if we should commit lock files for the examples, they might just create a lot of noise

rajbos replied:
Fair point, I'll remove them

rajbos replied:
Removed

gr2m commented on examples/1-hello-world/index.js:
```diff
@@ -0,0 +1,21 @@
+import { createServer } from "http";
```
Useing the [`node:` protocol](https://nodejs.org/api/esm.html#node-imports) is best practice

```suggestion
import { createServer } from "node:http";
```

It makes it clear that a module is not coming from npm

rajbos replied:
Done

gr2m commented on examples/1-hello-world/index.js:
```diff
@@ -0,0 +1,21 @@
+} from "@copilot-extensions/preview-sdk";
+
+const server = createServer((request, ressponse) => {
+  console.log(`Received [${request.method}]`);
```
I think we should log out both the method and the URL

```suggestion
  console.log(`Received ${request.method} ${request.url}`);
```

rajbos replied:
I see you removed the brackets as well. It's a preference of mine to be able to differentiate between fixed text and variable text, and makes it really easy to see when a value is empty.

gr2m replied:
we can leave in the brackets if you prefer. But values cannot be empty in this context

rajbos replied:
Added

gr2m reviewed (commented):
early feedback, I haven't looked at the handle prompt example yet.

I would suggest we remove `1-` prefixes, the names should stay on their own

rajbos commented:
> early feedback, I haven't looked at the handle prompt example yet.
> 
> I would suggest we remove `1-` prefixes, the names should stay on their own

I was aiming for having a very logical way of knowing where to start and gradually add more complex examples. Numbering them helps so that you do not need to search in the README.

gr2m commented on package.json:
```diff
@@ -21,6 +21,7 @@
   "license": "MIT",
   "description": "JavaScript SDK for Copilot Extensions",
   "dependencies": {
+    "@copilot-extensions/preview-sdk": "^4.0.3",
```
did you add this dependency on itself intentionally? If yes, why?

rajbos replied:
No, probably ran an npm install command at the wrong folder. Removed it!

rajbos reviewed (commented):
Resolved feedback comments in the code

gr2m commented on .gitignore:
```diff
@@ -1 +1,2 @@
 node_modules
+package-lock.json
```
we shouldn't ignore the lock file in the root folder, only in `examples/`

I think this should do:

```suggestion
/examples/**/package-lock.json
```

gr2m commented on examples/2-handle-prompt/index.js:
```diff
@@ -0,0 +1,100 @@
+import http from 'http';
```
```suggestion
import { createServer } from 'node:http';
```

gr2m commented on examples/2-handle-prompt/index.js:
```diff
@@ -0,0 +1,100 @@
+import http from 'http';
+import { Octokit } from "@octokit/rest";
```
Do not use `@octokit/rest`, use `octokit` instead. I know Copilot likes to recommend `@octokit/rest` because it's been around longer, but it's a legacy SDK. `octokit` is the official all-batteries-included SDK for GitHub's API now

gr2m commented on examples/2-handle-prompt/index.js:
```diff
@@ -0,0 +1,100 @@
+};
+
+// Create an HTTP server
+const server = http.createServer((req, res) => {
+    handler(req, res);
+});
```
```suggestion
const server = http.createServer(handler);
```

gr2m commented on examples/2-handle-prompt/index.js:
```diff
@@ -0,0 +1,100 @@
+
+// Define the handler function
+const handler = async (request, response) => {
+    if (request.method === 'POST') {
```
I would reverse condition and do an early return to avoid indentation

```suggestion
    if (request.method !== 'POST') {
```

gr2m commented on examples/2-handle-prompt/index.js:
```diff
@@ -0,0 +1,100 @@
+            try {                
+                // start sending the response back to Copilot
+                console.log('Handling response');
+                // header to indicate the state
+                response.writeHead(200, { 'Content-Type': 'application/json' });
```
I don't think this is needed. The content is not valid JSON code anyway

gr2m commented on examples/2-handle-prompt/index.js:
```diff
@@ -0,0 +1,100 @@
+                // this will also show the message "Copilot is responding" in the chat
+                response.write(createAckEvent());
+
+                // parse the incoming body as that has the information we need to handle the request / user prompt
+                const payload = parseRequestBody(body);
```
I think I would create a separate method to read the body text instead of moving all the core logic into the         `request.on('end', async () => {` handler. And in this example, we should verify the payload using https://github.com/copilot-extensions/preview-sdk.js/?tab=readme-ov-file#async-verifyrequestbykeyidrawbody-signature-keyid-requestoptions

gr2m commented on examples/2-handle-prompt/index.js:
```diff
@@ -0,0 +1,100 @@
+                // start writing text in the response
+                const textEvent = createTextEvent(`Hello ${userHandle}, \n\n`);
+                response.write(textEvent);
+                // add new lines to mark the difference between the fixed text and the dynamic text
+                response.write("\n\n");
```
the new lines are included in `textEvent`

```suggestion
```

gr2m commented on examples/2-handle-prompt/index.js:
```diff
@@ -0,0 +1,100 @@
+                // add new lines to mark the difference between the fixed text and the dynamic text
+                response.write("\n\n");
+                       
+                // add a first system prompt for the payload.messages to add instructions
+                payload.messages.unshift({ role: "system", content: "You are a helpful assistant that talks like a pirate." });	
```
WE are doing this system message unshifting inside `prompt` already

https://github.com/copilot-extensions/preview-sdk.js/blob/b7063d6f46af712f3ea884e409a5cb8dbe505a70/lib/prompt.js#L15C9-L25

```suggestion
```

gr2m commented on examples/2-handle-prompt/index.js:
```diff
@@ -0,0 +1,100 @@
+                const payload_message = payload.messages[payload.messages.length - 1];
+                const result = await prompt(payload_message.content, {
+                    messages: payload.messages, // we are giving the prompt the existing messages in this chat conversation
+                    model: "gpt-4",
```
setting `model` is optional, we default it to `gp-4o` right now

https://github.com/copilot-extensions/preview-sdk.js/blob/b7063d6f46af712f3ea884e409a5cb8dbe505a70/lib/prompt.js#L4

gr2m reviewed (commented):
sorry for all the comments. If you prefer, we can merge it as is and do a follow up pull request to discuss further, or leave out the 2nd example folder and do it in a separate pull request

rajbos commented:
@gr2m , I've addressed most of the feedback but moving the body parsing logic out of the `on_end` event. Please go over the comments and close them if the feedback has been handled adequately 😄.

gr2m commented on examples/2-handle-prompt/index.js:
```diff
@@ -0,0 +1,111 @@
+        response.writeHead(405, { 'Content-Type': 'text/plain' });
+        console.log(`Method ${request.method} not allowed`);
+        
+        response.end('Method Not Allowed');
+    }
```
we need to return here, otherwise the code below will get executed. That's what I meant with return early
```suggestion
        response.end('Method Not Allowed');
        return;
    }
```

gr2m commented on examples/2-handle-prompt/index.js:
```diff
@@ -0,0 +1,111 @@
+import { createServer } from 'node:http';
+import { Octokit } from "@octokit";
```
```suggestion
import { Octokit } from "octokit";
```

gr2m commented on examples/2-handle-prompt/index.js:
```diff
@@ -0,0 +1,111 @@
+    request.on('end', async () => {
+        try {
+            const payloadIsVerified = await verifyRequestByKeyId(
+                request.body,
```
```suggestion
                body,
```

gr2m commented on examples/2-handle-prompt/index.js:
```diff
@@ -0,0 +1,111 @@
+            const payload = parseRequestBody(body);
+            
+            // start writing text in the response
+            const textEvent = createTextEvent(`Hello ${userHandle}, \n\n`);
+            response.write(textEvent);
+            // add new lines to mark the difference between the fixed text and the dynamic text
+            response.write("\n\n");
```
can you explain what you mean by

> mark the difference between the fixed text and the dynamic text

None of the `\n` are necessary
```suggestion
            const textEvent = createTextEvent(`Hello ${userHandle}`);
            response.write(textEvent);
```

gr2m commented on examples/2-handle-prompt/index.js:
```diff
@@ -0,0 +1,111 @@
+            // the prompt to forward to the Copilot API is the last message in the payload
+            const payload_message = payload.messages[payload.messages.length - 1];
+            const result = await prompt(payload_message.content, {
+                system: "you talk like a pirate", // extra instructions for the prompt, the "you are a helpful assistant" is already set in the SDK
```
The "you are a helpful assistant" is just the default. If you want it to be part of your system prompt then you have to include it in your custom onew
```suggestion
                system: "You are a helpful assistant. You talk like a pirate",
```

gr2m commented on examples/2-handle-prompt/index.js:
```diff
@@ -0,0 +1,111 @@
+                token: tokenForUser         
+            });
+
+            // write the prompt response back to Copilot
+            // note that this is only send when the entire response from the Copilot API is ready
+            response.write(createTextEvent(result.message.content));
```
there is a `prompt.stream` API but we can do a separate example for that if you like?

gr2m commented on examples/2-handle-prompt/package.json:
```diff
@@ -0,0 +1,12 @@
+  },
+  "dependencies": {
+    "@copilot-extensions/preview-sdk": "latest",
+    "@octokit": "latest"
```
Don't use `latest` for external dependencies.

```suggestion
    "octokit": "^4.0.2"
```

gr2m reviewed (changes requested):
Let me know if you had a chance to test the apps once you applied all the remaining requested changes. If not I can do that too

gr2m commented:
We have a live stream today that I think might draw some interest, I think it would be great to finish this up by then:
https://nickyt.live/#2024-09-17-gregor-martynus%2c%20staff%20software%20engineer%20at%20github

I'll address my remarks to get the PR merged, we can do follow up PRs aftewards

gr2m reviewed (approved):
Thank you for getting an `examples/` folder started, this will be tremendously helpful. I should have started out with that 🙏🏼
